qiskit_experiments/warnings.py

```diff
@@ -122,3 +124,113 @@ def new(deprecated_cls, *args, **kwargs):
+        stacklevel: Stacklevel of this warning. See Python Warnings documentation for details.
+
+    Examples:
+
+        .. code-block::
+
+            @deprecated_argument(last_version="0.5", kwmap={"qubits": "physical_qubits"})
+            def function(*args, **kwargs):  # physical_qubits in in args or kwargs
+                pass
+
+    Returns:
+        Function with argument name replaced from old to new.
+    """
+
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            if kwargs:
+                _rename_kwargs(
+                    args[0].__class__.__name__ + "." + func.__name__,
+                    kwargs,
+                    kwarg_map,
+                    last_version,
+                    msg,
+                    stacklevel,
+                )
+            return func(*args, **kwargs)
+
+        return wrapper
+
+    return decorator
+
+
+def _rename_kwargs(func_name, kwargs, kwarg_map, last_version, msg, stacklevel):
+    for old_arg, new_arg in kwarg_map.items():
+        if old_arg in kwargs:
+            if new_arg in kwargs:
+                raise TypeError(f"{func_name} received both {new_arg} and {old_arg} (deprecated).")
+
+            message = f"{func_name} keyword argument {old_arg} is deprecated and will be removed "
+            if last_version:
+                message += f"after Qiskit Experiments {last_version}. "
+            else:
+                message += "in a future release. "
+            if new_arg is not None:
+                message += f"It is now replaced with {new_arg}. "
+                kwargs[new_arg] = kwargs.pop(old_arg)
+            if msg:
+                message += msg
+
+            warnings.warn(message, stacklevel=stacklevel)
+
+
+def qubit_deprecate() -> Callable:
+    """Decorator to deprecate from qubit to physcial_qubits"""
+
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            category = DeprecationWarning
+            func_name = args[0].__class__.__name__ + ".__init__"
+
+            if len(args) > 1 and isinstance(args[1], int):
+                args = list(args)
+                args[1] = [args[1]]
+                args = tuple(args)
+                warnings.warn(
+                    f"The first argument of {func_name} has been renamed from qubits to "
+                    "physical_qubits, and is expecting a sequence instead of an integer. "
+                    "Support of integer values is deprecated and will be removed after Qiskit "
+                    "Experiments 0.5.",
+                    category=category,
+                    stacklevel=3,
+                )
+
+            if kwargs and "qubit" in kwargs:
+                if "physical_qubits" in kwargs:
+                    raise TypeError(
+                        f"{func_name} received both physical_qubits and qubits (deprecated)."
```

Comment:
This is ambiguous—it could mean that supplying both physical_qubits and qubits is deprecated. I would rephrase to "received both physical_qubits and the deprecated qubits parameter, only physical_qubits should be supplied".